src/libfsm/union.c

```diff
@@ -157,231 +208,874 @@ fsm_union_array(size_t fsm_count,
+	while (state_set_next(&si, &s)) {
+		fprintf(f, " %d", s);
+	}
+	fprintf(f, "\n");
+}
+
+static void
+dump_edge_set(FILE *f, const char *name, fsm_state_t from, const struct edge_set *edges)
+{
+	struct edge_group_iter iter;
+	struct edge_group_iter_info info;
+	if (edge_set_empty(edges)) { return; }
+
+	fprintf(f, "  - %s:", name);
+	edge_set_group_iter_reset(edges, EDGE_GROUP_ITER_ALL, &iter);
+	while (edge_set_group_iter_next(&iter, &info)) {
+		fprintf(f, " %d->%d", from, info.to);
+	}
+	fprintf(f, "\n");
+}
+#endif
+
+/* For each state in the epsilon closure, if there's a labeled edge
+ * to an end state, check if the label set is only [\n] and there's
+ * also an epsilon edge to the same end state.
+ * If so, this represents an anchored end in the NFA. */
+static bool
+has_epsilon_and_newline_edges_to_same_end(const struct fsm *nfa, struct state_set *eclosure,
+    fsm_state_t s_i, fsm_state_t *dst_end)
+{
+	struct state_iter si;
+	state_set_reset(eclosure, &si);
+	fsm_state_t ns_i;
+	while (state_set_next(&si, &ns_i)) {
+		assert(ns_i < nfa->statecount);
+		const struct fsm_state *ns = &nfa->states[ns_i];
+
+		if (state_set_empty(ns->epsilons)) { continue; }
+		if (edge_set_empty(ns->edges)) { continue; }
+
+		struct edge_group_iter iter;
+		struct edge_group_iter_info info;
+		edge_set_group_iter_reset(ns->edges, EDGE_GROUP_ITER_ALL, &iter);
+		while (edge_set_group_iter_next(&iter, &info)) {
+			/* Look for an edge set with only '\n' */
+			if ((info.symbols[0] != (1ULL << '\n'))
+			    || info.symbols[1] || info.symbols[2] || info.symbols[3]) {
+				continue;
+			}
+
+			/* If it's an end, look for an epsilon leeding to the same destination */
+			if (fsm_isend(nfa, info.to)) {
+				struct state_iter inner_si;
+				fsm_state_t os_i;
+
+				assert(s_i < nfa->statecount);
+				const struct fsm_state *s = &nfa->states[s_i];
+
+				state_set_reset(s->epsilons, &inner_si);
+				while (state_set_next(&inner_si, &os_i)) {
+					if (os_i == info.to) {
+						*dst_end = info.to;
+						return true;
+					}
+				}
+			}
+		}
+	}
+
+	return false;
+}
+
+static bool
+has_labeled_edge_to_eclosure_with_unanchored_end_loop(const struct fsm *nfa,
+    struct state_set **eclosures,
+    fsm_state_t s_i, fsm_state_t unanchored_end_loop,
+    fsm_state_t *indirect_dst)
+{
+	if (unanchored_end_loop == NO_STATE) { return false; }
+	assert(unanchored_end_loop < nfa->statecount);
+
+	assert(s_i < nfa->statecount);
+	const struct state_set *s_eclosure = eclosures[s_i];
+
+	/* For every state in s_i's epsilon closure, check if it has
+	 * a labeled edge to a state with the unanchored_end_loop
+	 * in its epsilon closure. */
+	struct state_iter si;
+	state_set_reset(s_eclosure, &si);
+	fsm_state_t ns_i;
+	while (state_set_next(&si, &ns_i)) {
+		/* The unanchored_end_loop's self-edge doesn't count here. */
+		if (ns_i == unanchored_end_loop) { continue; }
+
+		/* FIXME: this should only apply to the original state, not its epsilon closure...right? */
```

Comment:
By construction the original state will be the first in its epsilon closure. I need to think through the big picture implications of this and possibly add some tests, but it should be irrelevant for our use case.